UW-494: Add page about supported tags to UW YAML docs (#463)
ufs-community/uwtools: docs/sections/user_guide/yaml/tags.rst

.. _defining_YAML_tags:

UW YAML tags
============

> Tags are used to denote the type of a YAML node when converting to a Python object. Standard YAML tags are defined at http://yaml.org/type/index.html.

Author: Thanks! A few other options felt awkward, so I went and used the `defined [here]`

Tags may be implicit:

.. code-block:: yaml

   boolean: true
   integer: 3
   float: 3.14

Or explicit:

.. code-block:: yaml

   boolean: !!bool "true"
   integer: !!int "3"
   float: !!float "3.14"

Additionally, UW defines the following tags to support use cases not covered by standard tags:

``!float``
^^^^^^^^^^

Converts the tagged node to a Python ``float`` value.

* Given input.yaml

>   .. code-block:: yaml
> 
>     f2: !float "{{ 3.141 + 2.718 }}"

Reviewer: ```suggestion
  .. code-block:: yaml

     f2: !float "{{ 3.141 + 2.718 }}"
```
nitpick: vertical alignment

Author: Oh the formatting all broke in there when I tried to add another level - I reverted , see if its ok now

  .. code-block:: text
     % uw config realize --input-file input.yaml --output-format yaml
     f2: 5.859

``!int``
^^^^^^^^

Converts the tagged node to a Python ``int`` value.

* Given input.yaml

>   .. code-block:: yaml
> 
>    values:
>     f1: 3
>     f2: 11
>     f3: !int "{{ (f1 + f2) * 10 }}"

Reviewer: ```suggestion
  .. code-block:: yaml

     values:
       f1: 3
       f2: 11
       f3: !int "{{ (f1 + f2) * 10 }}"
```

Author: Typo there along with the broken formatting! Should be ok now

.. _defining_YAML_tags:

UW YAML tags
============

Tags are used to denote the type of a YAML node when converting to a Python object. Standard YAML tags are defined at http://yaml.org/type/index.html.

Tags may be implicit:

.. code-block:: yaml

   boolean: true
   integer: 3
   float: 3.14

Or explicit:

.. code-block:: yaml

   boolean: !!bool "true"
   integer: !!int "3"
   float: !!float "3.14"

Additionally, UW defines the following tags to support use cases not covered by standard tags:

``!float``
^^^^^^^^^^

Converts the tagged node to a Python ``float`` value. For example:

> Given input.yaml:

Reviewer: ```suggestion
Given ``input.yaml``:
```
Only noticed when viewing the rendered HTML docs... Will add suggestions for the other cases.

> Converts the tagged node to a Python ``float`` value. For example:
> 
> Given ``input.yaml``:

Reviewer: ```suggestion
Converts the tagged node to a Python ``float`` value. For example, given ``input.yaml``:
```

Author: I debated the example and file on its line, or both separated, and ended up with this split to separate the ideas of the example from the description. But thats probably overkill, so this is a fair option!

.. code-block:: yaml

   f2: !float "{{ 3.141 + 2.718 }}"

.. code-block:: text

   % uw config realize --input-file input.yaml --output-format yaml
   f2: 5.859

``!int``
^^^^^^^^

Converts the tagged node to a Python ``int`` value. For example:

> Given input.yaml:

Reviewer: ```suggestion
Given ``input.yaml``:
```

> Converts the tagged node to a Python ``int`` value. For example:
> 
> Given ``input.yaml``:

Reviewer: ```suggestion
Converts the tagged node to a Python ``int`` value. For example, given ``input.yaml``:
```

.. code-block:: yaml

   f1: 3
   f2: 11
   f3: !int "{{ (f1 + f2) * 10 }}"

.. code-block:: text

   % uw config realize --input-file input.yaml --output-format yaml
   f1: 3
   f2: 11
   f2: 140

``!include``
^^^^^^^^^^^^

Parse the tagged file and include its tags. For example:

> Given input.yaml:

Reviewer: ```suggestion
Given ``input.yaml``:
```

> Parse the tagged file and include its tags. For example:
> 
> Given ``input.yaml``:

Reviewer: ```suggestion
Parse the tagged file and include its tags. For example, given ``input.yaml``:
```

.. code-block:: yaml

   values: !INCLUDE [./supplemental.yaml]

> and supplemental.yaml:

Reviewer: ```suggestion
and ``supplemental.yaml``:
```

.. code-block:: yaml

   e: 2.718
   pi: 3.141

.. code-block:: text

   % uw config realize --input-file input.yaml --output-format yaml
   values:
      e: 2.718
      pi: 3.141

``!remove``
^^^^^^^^^^^

Removes the tagged YAML key/value pair. For example:

> Given input.yaml:

Reviewer: ```suggestion
Given ``input.yaml``:
```

> Removes the tagged YAML key/value pair. For example:
> 
> Given ``input.yaml``:

Reviewer: ```suggestion
Removes the tagged YAML key/value pair. For example, given ``input.yaml``:
```

.. code-block:: yaml

   e: 2.718
   pi: 3.141

> and supplemental.yaml:

Reviewer: ```suggestion
and ``supplemental.yaml``:
```

.. _defining_YAML_tags:

> Defining YAML ``tags``
> ==========================

Reviewer: ```suggestion
UW YAML tags
============
```
The user is just using tags here, not defining them.

Tags are used to denote the type of a YAML node when converting to a Python object. Standard YAML tags are defined at http://yaml.org/type/index.html.

Tags may be implicit:

.. code-block:: yaml

   boolean: true
   integer: 3
   float: 3.14

Or explicit:

.. code-block:: yaml

   boolean: !!bool "true"
   integer: !!int "3"
   float: !!float "3.14"

> Additionally, UW extends the following tags and constructors to support other formats:

Reviewer: ```suggestion
Additionally, UW defines the following tags to support use cases not covered by standard tags:
```
I appreciate these docs and how they're making me think about what's really going on. For example, the `!!int`` standard tag doesn't work here:
```
% echo 'n: !!int "{{ 2 + 2 }}"' | uw config realize --input-format yaml --output-format yaml
...
ValueError: invalid literal for int() with base 10: '{{ 2 + 2 }}'
```
because `pyyaml` tries to convert the Jinja2 expression immediately, but our custom tag works:
```
% echo 'n: !int "{{ 2 + 2 }}"' | uw config realize --input-format yaml --output-format yaml
n: 4
```
because it only takes effect later, after Jinja2 has been used to render the expression. So my suggestion here to say "use cases" instead of "formats" is because the format is `int` either way, but the use case is different.

> 
> UW YAML Tags
> ------------

Reviewer: ```suggestion
```
Since this would be the page title already if an above suggestion is accepted.

``!float``
^^^^^^^^^^

> Converts the tagged node to a Python ``float`` type.

Reviewer: ```suggestion
Converts the tagged node to a Python ``float`` value.
```

.. code-block:: yaml

>    f2: !float '3.14'

Reviewer: ```suggestion
   f2: !float "{{ 3.141 + 2.718 }}"
```
Suggested because the current value could be handled by the built-in `!!float`, while the suggest value could not.

``!int``
^^^^^^^^

> Converts the tagged node to a Python ``integer`` type.

Reviewer: ```suggestion
Converts the tagged node to a Python ``int`` value.
```

.. code-block:: yaml

   cycle_day: !int "{{ cycle.strftime('%d') }}"

``!include``
^^^^^^^^^^^^

Parse the tagged file and include its tags.

.. code-block:: yaml

   salad: !INCLUDE [./fruit_config.yaml]

> ``!list``

Reviewer: Does this actually exist?

Author: AH - huh, now that you mention it, its in the ticket, but we never made a `!list` tag. As its optional in the AC, I'll remove it for now and we can talk about whether we actually need that

^^^^^^^^^

Converts the tagged node to a Python ``list`` type.

.. code-block:: yaml

   file_names: !list [&gfs_file_names]

``!remove``
^^^^^^^^^^^

Removes the tagged YAML key/value pair.

.. code-block:: yaml

>    update_values: !remove

Reviewer: What would you think about a fuller example, e.g.:

--

Given `input.yaml`
```
e: 2.718
pi: 3.141
```
and `supplemental.yaml`
```
e: !remove
```
```
% uw config realize --input-file input.yaml supplemental.yaml --output-format yaml
pi: 3.141
```

? If you like that, I think the other tags could likewise benefit from fuller examples showing inputs and outputs. The ones that involve only a single file could just read from stdin.

Author: Thats true! Honestly I found myself getting overly focused on things like 'is this how I want the title to look' and put it up to break that cycle - it can definitely do for the better examples, working on it!
